packages/server/graphql/public/mutations/deleteUsers.ts

```diff
@@ -0,0 +1,118 @@
+import {USER_BATCH_DELETE_LIMIT} from '../../../postgres/constants'
+import {getUsersByEmails} from '../../../postgres/queries/getUsersByEmails'
+import {getUserById} from '../../../postgres/queries/getUsersByIds'
+import {
+  getUserId,
+  isSuperUser,
+  isUserBillingLeader,
+  isUserOrgAdmin
+} from '../../../utils/authorization'
+import getDomainFromEmail from '../../../utils/getDomainFromEmail'
+import {GQLContext} from '../../graphql'
+import {markUserSoftDeleted} from '../../mutations/deleteUser'
+import softDeleteUser from '../../mutations/helpers/softDeleteUser'
+import {MutationResolvers} from '../resolverTypes'
+
+const deleteUsers: MutationResolvers['deleteUsers'] = async (
+  _source,
+  {emails}: {emails: string[]},
+  {authToken, dataLoader}: GQLContext
+) => {
+  if (emails.length === 0) {
+    return {error: {message: 'No any emails were provided'}}
+  }
+
+  if (emails.length > USER_BATCH_DELETE_LIMIT) {
+    return {error: {message: `Cannot delete more than ${USER_BATCH_DELETE_LIMIT} users at once`}}
+  }
+
+  const su = isSuperUser(authToken)
+  const viewerId = getUserId(authToken)
+  const viewer = await getUserById(viewerId)
+  if (!viewer) return {error: {message: 'Invalid viewer'}}
+
+  const usersToDelete = await getUsersByEmails(emails)
+
+  if (usersToDelete.length === 0) {
+    return {error: {message: 'No valid users found to delete'}}
+  }
+
+  const viewerOrgUsers = await dataLoader.get('organizationUsersByUserId').load(viewerId)
+  const viewerTeamMembers = await dataLoader.get('teamMembersByUserId').load(viewerId)
+
+  // Pre-compute org-level permissions
+  const viewerOrgPermissions = await Promise.all(
+    viewerOrgUsers.map(async (orgUser) => ({
+      orgId: orgUser.orgId,
+      isOrgAdmin: await isUserOrgAdmin(viewerId, orgUser.orgId, dataLoader),
+      isBillingLeader: await isUserBillingLeader(viewerId, orgUser.orgId, dataLoader)
+    }))
+  )
```

Comment:
+1 I find it not very easy to read the whole delete operation. I think the mutation would benefit from not trying to do these kind of optimizations, but instead do a simple
```
const deletedUsers = await Promise.all(usersToDelete.map(...
  // load objects and check permission
  // delete
).filter(isNotNull)
```
The dataloader can still partially still do it's thing and it's much easier to ensure we don't have any index errors.

If you want to go even further, we put the permission in a separate function
```
canDeleteUser(actor: User, user: User)
```

Reply:
I would like to implement the "all-or-nothing" strategy: if the caller does not have permissions to delete all of the users he/she requested, we won't delete any one and as a friendly robot, we return with error message to indicate which set of the users he/she requested cannot be deleted.
With that in mind, it's not possible to merge in the deletion operation into the permission check. WDYT?

Reply:
I see. I went back to the issue to see how this would look in practice. I'm not sure this would be the best user experience. What I expect a user to do in this case is searching the selected users to unselect the ones they can't delete and try again.
I'm also not sure deleting users is what is actually required for this issue.